spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/pubsub/PubSubHealthIndicatorAutoConfiguration.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2018-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.pubsub;
+
+import org.springframework.boot.actuate.autoconfigure.health.ConditionalOnEnabledHealthIndicator;
+import org.springframework.boot.actuate.health.HealthIndicator;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.cloud.gcp.pubsub.core.PubSubTemplate;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ *
+ * @author Vinicius Carvalho
+ *
+ * @since 1.2
+ */
+
+@Configuration
+@ConditionalOnClass(HealthIndicator.class)
+@ConditionalOnEnabledHealthIndicator("pubsub")
+@AutoConfigureAfter(GcpPubSubAutoConfiguration.class)
+public class PubSubHealthIndicatorAutoConfiguration {
```

Comment:
Why do we need to have this in the separate public class?
Why inner `static protected` one is not enough for our use-case?
Otherwise you need to add this top-level class also into the `spring.factories` file.

See `IntegrationAutoConfiguration` for example.

Reply:
@artembilan I'm following what the project convention seems to be doing, if you look into the DatastoreHealthIndicatorAutoConfiguration for example. Trying to be consistent to this project, glad to change if needed, but looks like this project uses a separate class convention. 
Good point on I'm missing `spring.factories` I totally forgot that one, resolved.

@meltsufin what do you prefer here on the style of the AutoConfiguration class?

Reply:
I think the choice has been done without consulting Spring Boot origins.
So, there is no consistency with what we are based on and you really have more work to do with that `spring.factories`.

But that's really IMHO and I can't make you do what I prefer 😄

Reply:
If we put the health indicator auto-configuration inside the top-level auto-configuration, wouldn't that mean that you can only use the health indicator when using the general auto-configuration as well, with say `spring.cloud.gcp.pubsub.enabled=true`? In other words, there's a behavior difference in using the inner config class. If that's the intended convention in Spring Boot, then sure, let's do that. I can update the Datastore one as well.

Reply:
Well, what is the point of health indicator for Pub/Sub if we disable such a sub-system in our application altogether?
Doesn't that mean we are just not interested in the health of a feature we have just disabled?

I just look at this from logical perspective.
Indeed we have its specific condition (`@ConditionalOnEnabledHealthIndicator("pubsub")`), but that's just fine-graining. If I use `spring.cloud.gcp.pubsub.enabled=false`, I'd feel like there should not be a health for Pub/Sub. 

I think we have been confused a bit because in Spring Boot health auto-configuration is done in the separate module - `spring-boot-actuator-autoconfigure`.
However it doesn't mean that we need to do a separate module here as well.

What I mean is confirmed with auto-configuration for Spring Cloud Gateway: https://github.com/spring-cloud/spring-cloud-gateway/blob/master/spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/config/GatewayAutoConfiguration.java

Any other opinion?

Reply:
Sure, makes sense for me. I don't feel strongly one way or another. If the convention seems to have the inner class, let's do that.